### ISSUE TITLE
[Bug] Tab history still visible after clearing all history

### ISSUE THREAD

**Issue by sv-ohorvath** [mozilla-mobile/fenix#13522]
## Steps to reproduce
1. Visit a few websites on a single tab.
2. Go to the History menu and delete all history. Or clear browsing data, leaving the Open tabs out.
3. Go back to the tab, long tap the back button.
### Expected behavior
No recent tab history should be displayed.
### Actual behavior
The tab's history is still displayed as long as you don't close the tab, even after an app restart.
### Device information

* Android device: any
* Fenix version: Beta 80.0.0-beta6, Nightly 8/11

**Comment by Mugurell** [comment]
Just to confirm with UX.
The described behavior is the same as for Fennec, where clearing history also deletes all tab history.
Upon deleting all history, all tabs would still be available, but with no history. Pressing back while on a tab would bring the user back to Fenix's HomeFragment.
Is this what we should implement?

**Comment by apbitner** [comment]
Pinging @brampitoyo who worked on the back history

**Comment by brampitoyo** [comment]
@Mugurell wrote:

1. …clearing history also deletes all tab history.
2. Upon deleting all history, all tabs would still be available, but with no history.
3. Pressing back while on a tab would bring the user back to Fenix's HomeFragment.

Yes. I totally agree on items 1 and 2. It’s the correct behaviour!

As for item 3, I think it would be consistent if the Back button is disabled, making the Plus icon the only way to access HomeFragment. Technically “the first page in tab history” is not HomeFragment. It’s actually nothing, because deleting all history would also delete “the first page”. Does that make sense, or is it confusing and would introduce more complication?

**Comment by Mugurell** [comment]
Thank you!

In regards to 
@brampitoyo wrote:
> As for item 3, I think it would be consistent if the Back button is disabled, making the Plus icon the only way to access HomeFragment. Technically “the first page in tab history” is not HomeFragment. It’s actually nothing, because deleting all history would also delete “the first page”. Does that make sense, or is it confusing and would introduce more complication?

I was thinking that indeed HomeFragment is not part of tab history but if one is to access a website (so it will have 1 item in history - the current page ) and then presses back she would get to the HomeFragment.
Same as with deleting tab history, the user would remain with just one history entry.. and then would press back, would get to HomeFragment.

This ^^ is what would happen now if we delete tab history.
If we are to disable the back button based on the "tab had history cleared" condition it would mean a bit of complexity.

**Comment by mobd** [comment]
I think the suggested change would go against users' expectations. If I see "clear browsing history" I would interpret that as the list you see if you view your history. I don't expect it to change any of my open tabs' state, otherwise I would have cleared open tabs as well.

I also don't think any other browser disables the back button for open tabs if the user clears his history. However, only disabling the tab history while still having the back button work would be inconsistent and possibly lead to confusion.

**Comment by liuche** [comment]
I think that getting 1 & 2 fixed would be the first priority, and let's do 3 as @Mugurell suggests, since the suggestion to disable back button will be very complicated, and the approach suggested here will be much better than the current state.

**Comment by brampitoyo** [comment]
@liuche @Mugurell Agreed. If disabling the back button is complicated, then let’s not do it.

**Comment by mcarare** [comment]
@brampitoyo Just to clarify:

> Expected behavior
> No recent tab history should be displayed

A newly opened tab with just an address inserted, no other navigation within tab, still shows an entry in the list, the current one. 
Should deleting all history make the list completely empty, or with just the one entry, the current one?

**Comment by brampitoyo** [comment]
@mcarare Deleting all history should make the list completely empty.

This means that, although the current tab may still show up on the screen, if you go to the History sub-page, the entry for that tab has been deleted and is no longer recorded.

If, from this tab, you tap on a link, then we will record the link you tap in the History. But the original link isn’t recorded anymore.

**Comment by mcarare** [comment]
@brampitoyo I was asking about the list that comes up when long pressing back button.

**Comment by brampitoyo** [comment]
@mcarare Yes. The tab history list (which appears when you long-press the Back icon) should also be erased together with the “main” history (which appears when you tap Main menu → History).

Because there’s no tab history, you can’t press the Back icon anymore.

**Comment by mcarare** [comment]
Waiting on https://github.com/mozilla-mobile/android-components/issues/4571.

**Comment by amedyne** [comment]
@mcarare Looks like the ac change got merged last Dec 8, can you check ?

**Comment by mcarare** [comment]
@amedyne The AC API implementation has landed, but it still needs work on Fenix side. Should we pull this in the current sprint?

**Comment by amedyne** [comment]
Yes, please pull in the work for Fenix to this sprint. Thanks!

**Comment by mcarare** [comment]
As a note for QA: Currently the "back history" will contain only the current tab, after history is cleared.
Regarding @brampitoyo specs, that the list should be empty, we encountered cases where the tab with reloading when being accessed again, and as such, the list will still be populated with the current tab. To eliminate inconsistencies, I have kept the current tab in the "back history" list.
@brampitoyo If UX would prefer to show the list empty on some tabs and with the reloaded tab on others, I can also implement those changes in a separate issue / PR.

**Comment by ebalazs-sv** [comment]
Verified as fixed on Nightly 1/13 with Nexus 5 (Android 6.0.1) and Pixel 2 (Android 9).
After clearing all history, the current tab is displayed when long pressing the device back button.

This is still reproducible on Beta 85.0.0-beta.7 from 1/12 and Release 84.1.4 from 1/7.

I will close this issue and remove the qa:needed label.

**Comment by ebalazs-sv** [comment]
This is still reproducible in Beta 85.0.0-beta.8 from 1/15 with Samsung Galaxy Tab S3 (Android 8).

**Comment by brampitoyo** [comment]
@mcarare wrote:

> …we encountered cases where the tab with reloading when being accessed again, and as such, the list will still be populated with the current tab. To eliminate inconsistencies, I have kept the current tab in the "back history" list.

I was working under the assumption that, even after clearing all history, the current page that’s displayed is still stored in the cache so it doesn’t need to be reloaded.

If the page is reloaded, then we can consider it a legitimate reason to “write an entry” into history. In this case, the fix you’ve done to eliminate inconsistencies is correct.

Hope this answers your questions? Let me know if I can clarify further.

**Comment by softvision-miralobontiu** [comment]
Reproducible on Beta 86.0.0-beta.5 with Sony Xperia Z5 Premium (Android 7.1.1).
